### DISCUSSION TITLE
Rules size limits, Berry language converter .

### DISCUSSION THREAD

**Discussion by jpier123** [arendst/Tasmota#19848] [General]
Dear Tasmota Community,

I was jut wondering what is the best aproach when taliking about more complex projects where three rules space is not enough. 

Could you please advice best aproach. Is there any Berry language converter that supports Tasmota rules  migration?

Thanks in advance,

JP

**Comment by sfromis** [comment]
I'm using [Berry](https://tasmota.github.io/docs/Berry/) on ESP32-based devices for such things, and have found it to work very well.

While rules can be fine for simple automations, Berry is much more powerful, and is a decent programming language, a bit like a simpler Python - but not available on ESP8266-based devices. There exists a [modification allowing up to 8 rule sets](https://github.com/arendst/Tasmota/pull/19438) instead of 3, but not merged into the project code. While providing relief for the size limit, it does not deal with the more basic issue of rules not being a good platform for complexity. 

The [Scripting](https://tasmota.github.io/docs/Scripting-Language/) feature is also available on ESP8266, and does allow more complexity if you get used to the concepts of how if works. Tends to be cryptic.

More generally, the original structure for Tasmota is that complex projects would be handled in a backend platform, controlling Tasmota devices via MQTT, instead of trying to do it all on the device itself. It would even be feasible to have an ESP32 with Tasmota and Berry manage automations on ESP8266 devices via MQTT.

And no, there is no language converter.
